feat: Spellcheck benchmark dataset and evaluation algorithm

jeremyarancio opened the pull request:
### What
Creation of the benchmark and evaluation algorithm to evaluate the spellcheck

### Benchmark
The benchmark is composed of 247 lists of ingredients from 3 data sources:
* 30% of the old dataset composed of manually corrected lists of ingredients in French from the previous work by Lucain W. Unmodified lists of ingredients are removed.
* 15 manually corrected lists of ingredients in different languages (used to prompt engineer OpenAI on the Spellcheck task)
* 100 lists of ingredients with the tag _50-percent-unknown_ corrected with GPT-3.5. It follows the correction guidelines.

### Argilla to validate benchmark
Lists of ingredients corrected with GPT-3.5 are checked and modified to respect the spellcheck guidelines.

### Evaluation algorithm
An evaluation algorithm is created to estimate the performance of the Spellcheck.
It calculates the Precision-Recall of the correction based on text sequences (Original-Reference-Prediction) by using tokenization and alignment algorithm.

jeremyarancio committed:
Spellcheck - Previous dataset processed

jeremyarancio committed:
Fix typos in Readme.md

jeremyarancio committed:
feat(spellcheck): :sparkles: Spellcheck architecture + Creation benchmark (WIP)

jeremyarancio committed:
feat(spellcheck): :sparkles: Prompt engineering + Benchmark creation

jeremyarancio committed:
Generate benchmark

jeremyarancio committed:
feat(spellcheck): :sparkles: Set up Argilla for benchmark correction + is_truncated

jeremyarancio committed:
feat(spellcheck): :sparkles: Validation benchmark created and verified with Argilla

jeremyarancio committed:
feat(spellcheck): :sparkles: Spellcheck evaluation algorithm developed

jeremyarancio committed:
Evaluation script works (save)

jeremyarancio committed:
Evaluation algorithm is working!

jeremyarancio committed:
docs(spellcheck): :memo: Finalize documentation evaluation algorithm

jeremyarancio committed:
Merge branch 'develop' into spellcheck_validation_data

jeremyarancio committed:
docs(spellcheck): :memo: Update README.md & correct code

jeremyarancio committed:
Fix README

jeremyarancio committed:
docs(spellcheck): :memo: Code documentation

jeremyarancio committed:
docs(spellcheck): :memo: Document evaluation algorithm & fix typos

jeremyarancio committed:
docs(spellcheck): :memo: Fix typos in readme

raphael0202 commented on spellcheck/README.md:
````diff
@@ -1,226 +1,182 @@
-```
-virtualenv -p /path/to/python3 .venv
-```
+* Correct typos in ingredient words if word is recognized;
````
```suggestion
* Correct typos in ingredient words if word is not recognized;
```

raphael0202 commented on spellcheck/README.md:
To me, data augmentation is quite different from what we do here: it's modifying the input data during training so that the model is robust to perturbations: https://en.wikipedia.org/wiki/Data_augmentation
In my opinion, the term "synthetic dataset generation" better describes what we're doing here.

raphael0202 commented on spellcheck/scripts/old_to_new/0_convert_old_data.py:
```diff
@@ -0,0 +1,92 @@
+
+def main():
+    convert_old_data(
+        orginal_path=ORIGINAL_DATA_PATH,
```
In all this file: `orginal` > `original`

raphael0202 commented on spellcheck/README.md:
The tokenization seems incorrect with the default tokenizer model, I have for original string:
`[1016, 8415, 4502, 389, 279, 282, 1425, 11]`

raphael0202 commented on spellcheck/README.md:
```suggestion
Original:       1016    8415   4502   389   279    282    1425   11
Reference:      791     8415   374    389   279    38681  13
Prediction:     1016    2466   8415   374   304    279    38681  13
```

jeremyarancio committed:
Reviews implemented

jeremyarancio committed:
Modify evaluation script to match readme

jeremyarancio committed:
Update Argilla benchmark with highlighted modif (WIP)

jeremyarancio committed:
feat(spellcheck): :sparkles: Update Benchmark Argilla dataset to introduce highlights

jeremyarancio committed:
Generate synthetic dataset (WIP)

jeremyarancio committed:
feat(spellcheck): :fire: Synthetic dataset generated & pushed to Argilla

jeremyarancio committed:
docs(spellcheck): :memo: Update guidelines + Doc conception training dataset

jeremyarancio committed:
fix(spellcheck): :ambulance: Evaluation algorithm improved

jeremyarancio committed:
fix(spellcheck): :ambulance: Update Readme to consider changes in Evaluation algorithm

jeremyarancio committed:
Update Readme

jeremyarancio committed:
Update Readme

jeremyarancio committed:
refactor(spellcheck): :art: Simplify structure of the spellcheck module

jeremyarancio committed:
Modify Evaluation module

jeremyarancio committed:
Evaluate gpt3.5 & Claude on Benchmark (WIP)

jeremyarancio committed:
feat(spellcheck): :zap: Evaluate LLMs on the benchmark using the Evaluation algorithm

LLMs evaluated: GPT-3.5-turbo, Claude-3-haiku - Claude-3-sonnet -Claude-3-opus

jeremyarancio committed:
Add LLM metrics

jeremyarancio committed:
fix(spellcheck): :sparkles: Extract benchmark-v2

jeremyarancio committed:
feat(spellcheck): :art: Evaluate LLMs

jeremyarancio committed:
LLM evaluation

jeremyarancio committed:
Small changes in readme

jeremyarancio committed:
Update guidelines

jeremyarancio committed:
fix(spellcheck): :bug: Enhance benchmark with new products

jeremyarancio committed:
fix(spellcheck): :bug: Benchmark v0.3 + Evaluation algo fixed + New evaluation

jeremyarancio committed:
Rename dataset files + correction errors in benchmark

jeremyarancio committed:
perf(spellcheck): :bookmark: Prompt update benchmark-v3.1

jeremyarancio committed:
refactor(spellcheck): :art: Change names in scripts after files renaming

jeremyarancio committed:
Prompt v3.2

jeremyarancio committed:
feat(spellcheck): :arrow_up: Benchmark-v4

No more correction of percentage - no more whitespace added if not necessary (if after punctuation for example)

jeremyarancio committed:
feat(spellcheck): :arrow_up: Prompt v3

jeremyarancio committed:
feat(spellcheck): :arrow_up: Prompt v4

jeremyarancio committed:
feat(spellcheck): :arrow_up: Prompt v5

jeremyarancio committed:
fix(spellcheck): :zap: Modify Correction Precision

Modify correction precision to consider all model true and false positives instead of only True Positives

jeremyarancio committed:
feat(spellcheck): :zap: Add normalization before evaluation: lowercase + whitespaces strip

jeremyarancio committed:
feat(spellcheck): :art: Add visual evaluation using Argilla

Argilla is used to evaluate the performance of the Spellcheck. Annotator can choose between good and bad correction, allowing us to focus on these errors.

jeremyarancio committed:
feat(spellcheck): :mag: Human evaluation GPT-3.5-turbo - benchmark v4 - pormpt v5

jeremyarancio committed:
feat(spellcheck): :art: Evaluation GPT-3.5-turbo - Benchmark v5 - Prompt v6

',' in lists added between ingrediients   /   Prompt modified to considered latest human eval

jeremyarancio committed:
feat(spellcheck): :art: Evaluation Spellcheck on perfectly parsed ingredients lists

jeremyarancio committed:
feat(spellcheck): 🎨 Evaluation Spellcheck on perfectly parsed ingredients lists (fix path error)

jeremyarancio committed:
refactor(spellcheck): :art: Refactoring + Unitests - part 1

jeremyarancio committed:
feat(spellcheck): :construction: Evaluate Gemini-1.0

jeremyarancio committed:
feat(spellcheck): 🚧 Evaluate Gemini-1.0

jeremyarancio committed:
feat(spellcheck): :zap: Gemini-1.5 Flash & Pro evaluated & New Correction Recall metric added

jeremyarancio committed:
feat(spellcheck): :hammer: Synthetic training dataset prepared and pushed to HF (v2)

jeremyarancio committed:
Flan-t5 training (WIP)

jeremyarancio committed:
Flan-T5 training and tracking working

jeremyarancio committed:
feat(spellcheck): :fire: Flan-T5 training script

jeremyarancio committed:
feat(spellcheck): :fire: Update metrics table with new champ

jeremyarancio committed:
feat(spellcheck): :fire: Training pipeline using Metaflow & Sagemaker is working

jeremyarancio committed:
feat(spellcheck): :art: Training flan-t5 with Sagemaker

jeremyarancio committed:
feat(spellcheck): :zap: Continous training

jeremyarancio committed:
Update Flan-T5 training

jeremyarancio committed:
Upload benchmark predictions to S3

jeremyarancio committed:
feat(spellcheck): :fire: Add human evaluation step in the training pipeline

- Evaluation is performed on the Sagemaker instance  - Evaluation data are sent to S3  - Metaflow human evaluation step get and push to Argilla

jeremyarancio committed:
fix(spellcheck): :bug: Fix Training Dataset Argilla module

Bug when preparing Argilla dataset

jeremyarancio committed:
fix(Spellcheck): :bug: Fix Human Evaluartion step: issue wrong S3 uri name

jeremyarancio committed:
fix(spellcheck): :zap: Add gradient checkpointing

jeremyarancio committed:
feat(spellcheck): :zap: Continue experimentation

jeremyarancio committed:
test(spellcheck): :construction: Add dotenv for local testing of the training script

jeremyarancio committed:
feat(spellcheck): :zap: Dataset v3

jeremyarancio committed:
feat(spellcheck): :art: Deploy Mistral 7B on Inference Endpoints and implementation into code

jeremyarancio committed:
feat(spellcheck): :art: Evaluation Mistral-7B on benchmark

raphael0202 committed:
Merge branch 'develop' into spellcheck_validation_data
